README.rst
.. contents::

.. image:: https://travis-ci.org/plone/plone.recipe.codeanalysis.png?branch=master
    :target: http://travis-ci.org/plone/plone.recipe.codeanalysis

.. image:: https://coveralls.io/repos/plone/plone.recipe.codeanalysis/badge.png?branch=master
    :target: https://coveralls.io/r/plone/plone.recipe.codeanalysis

.. image:: https://pypip.in/d/plone.recipe.codeanalysis/badge.png
        :target: https://crate.io/packages/plone.recipe.codeanalysis

.. image:: https://pypip.in/v/plone.recipe.codeanalysis/badge.png
        :target: https://crate.io/packages/plone.recipe.codeanalysis


Introduction
============

``plone.recipe.codeanalysis`` provides static code analysis for Buildout-based
Python projects, including `flake8`_, `JSHint`_, `CSS Lint`_, and
other code checks.

This buildout recipe creates a script to run the code analysis::

    bin/code-analysis

By default ``plone.recipe.codeanalysis`` also creates a git pre-commit hook, in
order to run the code analysis automatically before each commit.

``plone.recipe.codeanalysis`` comes with a Jenkins integration, that allows to use
the same code analysis settings on your local machine as well as on Jenkins.


Installation
============

Just add a code-analysis section to your buildout.cfg::

    [buildout]
    parts += code-analysis

    [code-analysis]
    recipe = plone.recipe.codeanalysis
    directory = ${buildout:directory}/src

The directory option is not required. Though, if you don't specify a directory
the code analysis will check every file in your buildout directory.


Jenkins Installation
====================

plone.recipe.codeanalysis provides a Jenkins setting that allows to run it on a Jenkins CI server and to process and integrate the output via the
`Jenkins Violations plugin`_.

Usually you don't want the recipe to create Jenkins output files on your
local machine. Therefore it makes sense to enable the Jenkins output only
on the CI machine. To do so, just create a jenkins.cfg that extends and
overrides the default buildout file (that includes the other settings)::

    [buildout]
    parts += code-analysis

    [code-analysis]
    recipe = plone.recipe.codeanalysis
    jenkins = True

The Jenkins job itself should run "bin/code-analysis"::

    python bootstrap.py -c jenkins.cfg
    bin/buildout -c jenkins.cfg
    bin/jenkins-test --all
    bin/code-analysis

The `Jenkins Violations plugin`_ needs to be configured to read the output
files generated by this configuration.

pep8 (to read the flake8 output)::

    **/parts/code-analysis/flake8.log

csslint::

    **/parts/code-analysis/csslint.xml

jslint (to read the jshint output::

    **/parts/code-analysis/jshint.xml

Filesystem output
=================

If jenkins is set to False, you can still store the output on the filesystem by setting flake8-filesystem=True.
This is ignored if jenkins is set to True.

output::

    **/parts/code-analysis/flake8.txt

Links
=====

Code repository:

    https://github.com/plone/plone.recipe.codeanalysis

Continuous Integration:

    https://travis-ci.org/plone/plone.recipe.codeanalysis

Issue Tracker:

    https://github.com/plone/plone.recipe.codeanalysis/issues


Supported options
=================

The recipe supports the following options:

**directory**
    Directory that is subject to the code analysis.

**pre-commit-hook**
    If set to True, a git pre-commit hook is installed that runs the code
    analysis before each commit. Default is ``True``.

**jenkins**
    If set to True, the flake8, jshint and csslint code analysis steps will
    write output files that can be processed by the
    `Jenkins Violations plugin`_. Default is ``False``.

**flake8-filesystem**
    If set to True, the flake8 code analysis step will
    write an output file. Ignored if jenkins is True. Default is ``False``.

**flake8**
    If set to True, run Flake8 code analysis. Default is ``True``.

**flake8-ignore**
    Skip errors or warnings. See `Flake8 documentation`_ for error codes.
    Default is none.

**flake8-exclude**
    Comma-separated filename and glob patterns default. Say you want to
    exclude bootstrap.py, setup.py and all collective.* and plone.* packages.
    Just set "flake8-exclude=bootstrap.py,docs,*.egg,setup.py,collective.*,
    plone.*" in your buildout configuration. Default is
    ``bootstrap.py,docs,*.egg``.

**flake8-max-complexity**
    McCabe complexity threshold. Default is ``10``.

**flake8-max-line-length**
    Set maximum allowed line length. Default is ``79``.

**jshint**
    If set to True, jshint code analysis is run. Default is ``False``. Note
    that plone.recipe.codeanalysis requires jshint >= 1.0.

**jshint-bin**
    JSHint executable. Default is ``jshint``. If you have JSHint installed on
    your system and in your path, there is nothing to do. To install JSHint in
    your buildout, use the following::

        [jshint]
        recipe = gp.recipe.node
        npms = jshint
        scripts = jshint

    set jshint-bin to '${buildout:directory}/bin/jshint'.

**jshint-exclude**
    Allows you to specify directories which you don't want to be linted.
    Default is none. If you want JSHint to skip some files you can list them
    in a file named ``.jshintignore``. See `JSHint documentation`_ for more
    details.

**jscs**
    If set to True, jscs code analysis is run. Default is ``False``.

    JavaScript Code Style options should be configured using a ``.jscs.json``
    file. You should align your javascript code to the next generation of
    Plone's javascript framework Mockup_ and take it's ``.jscs.json`` file
    which is available here:
    https://github.com/plone/mockup/blob/master/.jscs.json

    All configuration options are documented on the `jscs website`_.

**jscs-bin**
    Set the path to a custom version of JSCS, e.g.
    "/usr/local/bin/jscs".

    If you have Javascript Code Style Checker installed in your system and
    path, you have nothing to do. To install with Buildout, add the following
    section to your buildout and set jscs-bin to
    ``{buildout:bin-directory}/jscs``::

        [jscs]
        recipe = gp.recipe.node
        npms = jscs
        scripts = jscs

**jscs-exclude**
    Allows you to specify directories and/or files which you don't want to be
    checked. Default is none. Note that these directories have to be given in
    absolute paths, use ``${buildout:directory}/foo/bar/static/js-3rd-party``
    for example.

**csslint**
    If set to True, CSS Lint code analysis is run. Default is ``False``.

    CSS Lint options should be configured using a ``.csslintrc`` file. A
    typical ``.csslintrc`` file will look like this::

        --format=compact
        --quiet
        --ignore=adjoining-classes,floats,font-faces,font-sizes,ids,qualified-headings,unique-headings
        --exclude-list=foo/bar/static/third-party.css

    This typical configuration includes a list of CSS rules that will be
    ignored as they are `considered useless`_.

    See `CSS Lint documentation`_ for a detailed list and description of the
    rules.

**csslint-bin**
    Set the path to a custom version of CSS Lint, e.g.
    "/usr/local/bin/csslint".

    If you have CSS Lint installed in your system and path, you have nothing
    to do. To install CSS Lint with Buildout, add the following section to
    your buildout and set csslint-bin to
    ``{buildout:bin-directory}/csslint``::

        [csslint]
        recipe = gp.recipe.node
        npms = csslint
        scripts = csslint

**deprecated-aliases**
    For historical reasons, some of the unittest.TestCase methods had one or
    more aliases that are deprecated on Python 2.7. If this option is set to
    True, warnings about deprecated aliases will be printed. Default is
    ``False``. See `Unit testing framework documentation`_ for more
    information.

**utf8-header**
    If set to True, Python files without a utf-8 header (like
    ``# -*- coding: utf-8 -*-``) will cause a warning. Default is ``False``.

**clean-lines**
    If set to True, **any file** containing trailing spaces or tabs anywhere
    on the lines will cause a warning. Default is ``False``.

**clean-lines-exclude**
    Allows you to specify directories and/or files which you don't want to be
    checked. Default is none. Note that these directories have to be given in
    absolute paths, use ``${buildout:directory}/foo/bar/static/js-3rd-party``
    for example.

**prefer-single-quotes**
    If set to True, Python files will be scanned searching for strings quoted
    with double quote signs (``"``). Default is ``False``.

**pep3101**
    If set to True, Python files will be scanned in search of existing '%'
    string formatting operators. Default is ``False``. See `PEP 3101 (Advanced
    String Formatting)`_ for more information.

**imports**
    If set to True, checks that imports in Python files follow `plone.api
    conventions`_. This also includes checking for alphabetically sorted
    import statements. Default is ``False``.

**debug-statements**
    If set to True, scan Python files looking for debug-like statements.
    Default is ``False``.

**return-status-codes**
    If set to True, the ``bin/code-analysis`` script returns an error code
    that Continuous Integration servers (like Travis CI) can use to fail or
    pass a job, based on the code analyis output. Note that Jenkins usually
    does not need this option (this is better handled by the Jenkins
    Violations plugin). Note that this option does not have any effect on the
    other code analysis scripts. Default is ``False``.

**rst-spelling**
    If set tot True, all ``.rst`` files will be checked for spelling mistakes.

i18ndude and zptlint support
----------------------------

To reduce the number of Zope/Plone direct dependencies, plone.recipe.codeanalysis no longer depends on `i18ndude`_ nor `zptlint`_;
in order to use the following options you have to install them on your
system:

**find-untranslated**
    If set to True, scan Zope templates to find untranslated strings.
    Default is ``False``.
    To use this you will need to set the ``i18ndude-bin`` option.

**i18ndude-bin**
    Set the path to a custom version of `i18ndude`_.
    Default is none.

**zptlint**
    If set to True, zptlint code analysis is run.
    Default is ``False``.
    To use this you will need to set the ``zptlint-bin`` option.

**zptlint-bin**
    Set the path to a custom version of `zptlint`_.
    Default is none.


Known Issues
============

JSHint "ERROR: Unknown option --verbose"::

    JSHint                [ OK ]
    ERROR: Unknown option --verbose

Upgrade JSHint to latest version (>= 1.0) to fix this issue, e.g.::

    $ sudo npm install -g jshint


JSHint "ERROR: Unknown option --exclude"::

    JSHint                [ OK ]
    ERROR: Unknown option --exclude

Upgrade JSHint to latest version (>= 2.1.6) to fix this issue, e.g.::

    $ sudo npm install -g jshint


.. _`considered useless`: http://2002-2012.mattwilcox.net/archive/entry/id/1054/
.. _`CSS Lint documentation`: https://github.com/stubbornella/csslint/wiki/Rules
.. _`CSS Lint`: http://csslint.net/
.. _`Flake8 documentation`: http://flake8.readthedocs.org/en/latest/warnings.html#error-codes
.. _`Jenkins Violations plugin`: https://wiki.jenkins-ci.org/display/JENKINS/Violations
.. _`flake8`: https://pypi.python.org/pypi/flake8
.. _`JSHint documentation`: http://jshint.com/docs/
.. _`JSHint`: http://www.jshint.com/
.. _`PEP 3101 (Advanced String Formatting)`: http://www.python.org/dev/peps/pep-3101/
.. _`plone.api conventions`: http://ploneapi.readthedocs.org/en/latest/contribute/conventions.html#about-imports
.. _`zptlint`: https://pypi.python.org/pypi/zptlint
.. _`i18ndude`: https://pypi.python.org/pypi/i18ndude
.. _`Unit testing framework documentation`: http://docs.python.org/2/library/unittest.html#deprecated-aliases
.. _`Mockup`: https://github.com/plone/mockup
.. _`jscs website`: https://www.npmjs.org/package/jscs


CHANGES.rst
Change history
==============

1.0rc2 (unreleased)
-------------------

- Add rst spelling checker.
  [gforcada]

- Add option to store flake8 output if jenkins is False
  [Michael Davis]


1.0rc1 (2014-06-18)
-------------------

- Return a string to avoid TypeError when no file was checked with ``jscs``.
  [saily]

- Check import sorting in ``code_analysis_imports`` and add tests for
  clean and sorted imports.
  [saily]

- Refactor ``code_analysis_clean_lines`` to use a new method to retrieve
  files and avoid too complex violation.
  [saily]


1.0b8 (2014-06-05)
------------------

- Add ``clean-lines-exclude`` support and updated README.
  [saily]

- Added tests for clean-lines checks.
  [saily]

- Use indices for format() to support Python 2.6. This fixes #77.
  [timo]


1.0b7 (2014-05-04)
------------------

- Add Javascript Code Style Checker ``jscs`` support.
  [saily]

- Remove hard dependency on i18ndude and zptlint; this will reduce the number
  of Zope/Plone direct dependencies to make life happier to people using
  Pyramid and other web Python-based development frameworks (closes `#53`_).
  [hvelarde]

- Do not print out jshint and csslint output for Jenkins. Those files can
  become quite large.
  [timo]


1.0b6 (2013-10-16)
------------------

- Remove progress bullets from flake8 check.
  [timo]

- Improve the way to handle an exception if the command used in popen does
  not exist.
  [flohcim]


1.0b5 (2013-10-08)
------------------

- Fix code analysis method by making it call each check only if the option
  is activated.
  [flohcim]

- Keep backward compatibility with 'string-formatting' option.
  [hvelarde]

- Rename 'deprecated-alias' to 'deprecated-aliases' and keep backward
  compatibility.
  [hvelarde]


1.0b4 (2013-10-06)
------------------

- Implement Jenkins option on CSS Lint and JSHint.
  [hvelarde, ramiroluz]

- Rename 'deprecated-methods' to 'deprecated-alias'.
  [gforcada]

- Rename 'string-formatting' option to 'pep3101' to keep consistency.
  [hvelarde]

- Remove unused CSSLINT_IGNORE remainings.
  [timo]

- Simplify code analysis method and make it more readable.
  [timo]


1.0b3 (2013-09-12)
------------------

- Add return-status-codes option that allows to fail a CI-build on Travis.
  [timo]

- Make system wide installed csslint the default value for
  the csslint-bin option.
  [timo]


1.0b2 (2013-09-11)
------------------

- Deprecate 'csslint-quiet', 'csslint-ignore' and 'csslint-exclude-list'
  options; CSS Lint must be configured now using a '.csslintrc' file.
  'csslint-bin' option now defaults to ``bin/csslint``; documentation was
  updated (closes #20).
  [hvelarde]

- Implement removal of pre-commit hook (fixes #21).
  [hvelarde]


1.0b1 (2013-08-12)
------------------

- Workaround over JSHint limitations to avoid displaying warning messages as
  errors (closes #13).
  [hvelarde]

- Fix CSS Lint validation and implement new 'csslint-quiet' option.
  [hvelarde]

- Fix package distribution.
  [hvelarde]


1.0a1 (2013-08-04)
------------------

- Initial release.
  [timo]

.. _`#53`: https://github.com/plone/plone.recipe.codeanalysis/issues/53


plone/recipe/codeanalysis/tests/test_jshint.py
# -*- coding: utf-8 -*-
import unittest
from plone.recipe.codeanalysis.jshint import code_analysis_jshint
# from plone.recipe.codeanalysis.jshint import run_cmd
from plone.recipe.codeanalysis.jshint import jshint_errors
from shutil import rmtree
from tempfile import mkdtemp
from os.path import join as path_join
from os.path import isfile as path_isfile


class TestJSHint(unittest.TestCase):
    def setUp(self):
        self.options = {
            'jshint-bin': 'bin/jshint',
            'jshint-exclude': '',
            'jenkins': 'False'
        }
        self.test_dir = mkdtemp()

    def tearDown(self):
        rmtree(self.test_dir)

    def test_analysis_should_return_false_when_error_found(self):
        incorrect_code = file(path_join(self.test_dir, 'incorrect.js'), 'w')
        incorrect_code.write(
            'var number_ten= =10;'
            'var word_ten=\'ten\';'
            'var sum_2_plus_2 = 2+2;')
        incorrect_code.close()
        self.options['directory'] = self.test_dir
        self.assertFalse(code_analysis_jshint(self.options))

    # def test_analysis_should_output_warnings(self):
    #     warnings_code = file(path_join(self.test_dir, 'warnings.js'), 'w')
    #     warnings_code.write(
    #         'function slideJump() {'
    #         '    if (window.location.hash == null || '
    #         'window.location.hash == \'\') {'
    #         '        return;'
    #         '    }'
    #         '    if (window.location.hash == null) return;'
    #         '    if (dest == null) {'
    #         '        dest = 0;'
    #         '    }'
    #         '}')
    #     warnings_code.close()
    #     self.options['directory'] = self.test_dir
    #     expected_output = \
    #         '{0[directory]}/warnings.js: line 1, col 52, Use \'===\' to' \
    #         ' compare with \'null\'. (W041)\n{0[directory]}/warnings.js:' \
    #         ' line 1, col 84, Use \'===\' to compare with \'\'. (W041)\n' \
    #         '{0[directory]}/warnings.js: line 1, col 141, Use \'===\' to' \
    #         ' compare with \'null\'. (W041)\n{0[directory]}/warnings.js:' \
  #         ' line 1, col 170, Use \'===\' to compare with \'null\'. (W041)' \
    #         '\n\n4 errors\n'.format(self.options)
    #     output = run_cmd(self.options, False)
    #     self.assertEqual(output, expected_output)

    # def test_analysis_should_return_true_for_warnings(self):
    #     warnings_code = file(path_join(self.test_dir, 'warnings.js'), 'w')
    #     warnings_code.write(
    #         'function slideJump() {'
    #         '    if (window.location.hash == null || '
    #         'window.location.hash == \'\') {'
    #         '        return;'
    #         '    }'
    #         '    if (window.location.hash == null) return;'
    #         '    if (dest == null) {'
    #         '        dest = 0;'
    #         '    }'
    #         '}')
    #     warnings_code.close()
    #     self.options['directory'] = self.test_dir
    #     self.assertTrue(code_analysis_jshint(self.options))

    def test_analysis_should_return_false_when_oserror(self):
        # The options are fake, so the function should raise an OSError
        # and return false.
        self.options['jshint-bin'] = 'FAKE_BIN'
        self.options['directory'] = 'FAKE_DIR'
        self.assertFalse(code_analysis_jshint(self.options))

    # def test_analysis_should_return_true(self):
    #     correct_code = file(path_join(self.test_dir, 'correct.js'), 'w')
    #     correct_code.write(
    #         'var number_ten=10;'
    #         'var word_ten=\'ten\';'
    #         'var sum_2_plus_2 = 2+2;')
    #     correct_code.close()
    #     self.options['directory'] = self.test_dir
    #     self.assertTrue(code_analysis_jshint(self.options))

    def test_analysis_file_should_exist_when_jenkins_is_true(self):
        location_tmp_dir = mkdtemp()
        correct_code = file(path_join(self.test_dir, 'correct.js'), 'w')
        correct_code.write(
            'var number_ten=10;'
            'var word_ten=\'ten\';'
            'var sum_2_plus_2 = 2+2;')
        correct_code.close()
        self.options['directory'] = self.test_dir
        self.options['location'] = location_tmp_dir
        self.options['jenkins'] = 'True'  # need to activate jenkins.
        code_analysis_jshint(self.options)
        file_exist = path_isfile(path_join(location_tmp_dir, 'jshint.xml'))
        rmtree(location_tmp_dir)
        self.assertTrue(file_exist)

    def test_jshint_errors_should_return_false_xml_output(self):
        output = '<?xml version="1.0" encoding="utf-8"?>\n' \
                 '<jslint>\n' \
                 '</jslint>'
        self.assertFalse(jshint_errors(output, True))

    def test_jshint_errors_should_return_true_xml_output(self):
        output = '<?xml version="1.0" encoding="utf-8"?>\n' \
            '<jslint>\n' \
            '    <file name="incorrect.js">\n' \
            '        <issue line="1" char="17" reason="Expected an ' \
            'identifier and instead saw &apos;=&apos;." evidence="var ' \
            'number_ten= =10;var word_ten=&apos;ten&apos;;var ' \
            'sum_2_plus_2 = 2+2;" severity="E" />\n' \
            '        <issue line="1" char="18" reason="Missing semicolon." ' \
            'evidence="var number_ten= =10;var word_ten=&apos;ten&apos;;var' \
            ' sum_2_plus_2 = 2+2;" severity="W" />\n' \
            '        <issue line="1" char="18" reason="Expected an ' \
            'assignment or function call and instead saw an expression." '\
            'evidence="var number_ten= =10;var word_ten=&apos;ten&apos;;var' \
            ' sum_2_plus_2 = 2+2;" severity="W" />\n' \
            '    </file>\n' \
            '</jslint>'
        self.assertTrue(jshint_errors(output, True))

    def test_jshint_errors_should_return_true_normal_output(self):
        output = 'incorrect.js: line 1, col 17, Expected an identifier and ' \
            'instead saw \'=\'. (E030)\nincorrect.js: line 1, col 18, ' \
            'Missing semicolon. (W033)\nincorrect.js: line 1, col 18, ' \
            'Expected an assignment or function call and instead saw an ' \
            'expression. (W030)\n\n3 errors\n'
        self.assertTrue(jshint_errors(output, False))

    def test_jshint_errors_should_return_false_normal_output(self):
        output = ''
        self.assertFalse(jshint_errors(output, False))


plone/recipe/codeanalysis/rst.py
# -*- coding: utf-8 -*-
from plone.recipe.codeanalysis.utils import find_files
from plone.recipe.codeanalysis.utils import log
from sphinxcontrib.spelling.checker import SpellingChecker

import os


def code_analysis_rst_spelling(options):
    log('title', 'rst spelling')

    found_files = find_files(options, '.*\.rst')
    file_paths = set(found_files.strip().split('\n'))
    word_list = os.path.join(os.path.dirname(__file__), 'rst_word_list.txt')

    checker = SpellingChecker(
        lang='en_US',
        suggest=True,
        word_list_filename=word_list,
    )

    all_typos = []
    for file_path in file_paths:
        with open(file_path) as f:
            typos = process_file(f, file_path, checker)
            if typos:
                all_typos += typos  # keep all typos in a flat list

    if len(all_typos) > 0:
        log('failure')
        for typo in all_typos:
            print('{0}:{1} {2} {3}'.format(*typo))
        print('Total errors: {0}'.format(len(all_typos)))
        return False
    else:
        log('ok')
        return True


def process_file(f, file_path, checker):
    typos = []
    for line_number, line_text in enumerate(f):
        if '>>>' in line_text:
            continue
        if '...' in line_text:
            continue
        for word, suggestions in checker.check(line_text):
            typos.append([file_path, line_number + 1, word, suggestions])
    return typos


plone/recipe/codeanalysis/__init__.py
# -*- coding: utf-8 -*-
"""Recipe codeanalysis"""
from plone.recipe.codeanalysis.clean_lines import code_analysis_clean_lines
from plone.recipe.codeanalysis.csslint import code_analysis_csslint
from plone.recipe.codeanalysis.debug_statements import \
    code_analysis_debug_statements
from plone.recipe.codeanalysis.deprecated_aliases import \
    code_analysis_deprecated_aliases
from plone.recipe.codeanalysis.flake8 import code_analysis_flake8
from plone.recipe.codeanalysis.i18ndude import code_analysis_find_untranslated
from plone.recipe.codeanalysis.imports import code_analysis_imports
from plone.recipe.codeanalysis.jscs import code_analysis_jscs
from plone.recipe.codeanalysis.jshint import code_analysis_jshint
from plone.recipe.codeanalysis.pep3101 import code_analysis_pep3101
from plone.recipe.codeanalysis.python_utf8_header import \
    code_analysis_utf8_header
from plone.recipe.codeanalysis.quoting import \
    code_analysis_prefer_single_quotes
from plone.recipe.codeanalysis.rst import code_analysis_rst_spelling
from plone.recipe.codeanalysis.zptlint import code_analysis_zptlint
import os
import subprocess
import zc.buildout
import zc.recipe.egg


current_dir = os.path.dirname(__file__)


class Recipe(object):
    """zc.buildout recipe"""

    def __init__(self, buildout, name, options):
        self.buildout, self.name, self.options = buildout, name, options
        self.egg = zc.recipe.egg.Scripts(
            buildout,
            self.options['recipe'],
            options
        )

        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'], name)

        # Set required default options
        self.options.setdefault('directory', '.')
        self.options.setdefault('pre-commit-hook', 'True')
        # Flake 8
        self.options.setdefault('flake8', 'True')
        self.options.setdefault('flake8-ignore', '')
        self.options.setdefault('flake8-exclude', 'bootstrap.py,docs,*.egg')
        self.options.setdefault('flake8-max-complexity', '10')
        self.options.setdefault('flake8-max-line-length', '79')
        # JSHint
        self.options.setdefault('jshint', 'False')
        self.options.setdefault('jshint-bin', 'jshint')
        self.options.setdefault('jshint-exclude', '')
        # JSCS
        self.options.setdefault('jscs', 'False')
        self.options.setdefault('jscs-bin', 'jscs')
        self.options.setdefault('jscs-exclude', '')
        # CSS Lint
        self.options.setdefault('csslint', 'False')
        self.options.setdefault('csslint-bin', 'csslint')
        # Warn about usage of deprecated aliases
        self.options.setdefault('deprecated-aliases', 'False')
        # XXX: keep compatibility with previous versions
        if self.options['deprecated-aliases'] == 'False':
            self.options.setdefault('deprecated-alias', 'False')
            deprecated_alias = self.options['deprecated-alias']
            if deprecated_alias == 'False':
                self.options.setdefault('deprecated-methods', 'False')
                deprecated_methods = self.options['deprecated-methods']
                if deprecated_methods != 'False':
                    self.options['deprecated-aliases'] = deprecated_methods
            else:
                self.options['deprecated-aliases'] = deprecated_alias
        # utf-8 header
        self.options.setdefault('utf8-header', 'False')
        # clean lines
        self.options.setdefault('clean-lines', 'False')
        self.options.setdefault('clean-lines-exclude', '')
        # Prefer single quotes over double quotes
        self.options.setdefault('prefer-single-quotes', 'False')
        # PEP 3101 (Advanced String Formatting)
        self.options.setdefault('pep3101', 'False')
        # XXX: keep compatibility with previous versions
        if self.options['pep3101'] == 'False':
            self.options.setdefault('string-formatting', 'False')
            if self.options['string-formatting'] != 'False':
                self.options['pep3101'] = self.options['string-formatting']
        # imports
        self.options.setdefault('imports', 'False')
        # Debug statements
        self.options.setdefault('debug-statements', 'False')
        # Jenkins output
        self.options.setdefault('jenkins', 'False')
        self.options.setdefault('flake8-filesystem', 'False')
        # Error codes
        self.options.setdefault('return-status-codes', 'False')
        # rst spelling
        self.options.setdefault('rst-spelling', 'False')
        # Find untranslated strings
        self.options.setdefault('find-untranslated', 'False')
        self.options.setdefault('i18ndude-bin', '')
        # zptlint
        self.options.setdefault('zptlint', 'False')
        self.options.setdefault('zptlint-bin', '')
        # Figure out default output file
        plone_jenkins = os.path.join(
            self.buildout['buildout']['parts-directory'], 'code-analysis'
        )
        if not os.path.exists(plone_jenkins):
            os.makedirs(plone_jenkins)

        # What files are tracked by this recipe
        self.files = [
            plone_jenkins,
            os.path.join(
                self.buildout['buildout']['bin-directory'], self.name
            )
        ]

    def install(self):
        self.install_scripts()

        # XXX: this has to be handled on a better way; what about 'false'?
        if self.options['pre-commit-hook'] != 'False':
            self.install_pre_commit_hook()
        else:
            self.uninstall_pre_commit_hook()

        # Create location
        wd = self.options.get('working-directory', '')
        if not wd:
            wd = self.options['location']
            if os.path.exists(wd):
                assert os.path.isdir(wd)
            else:
                os.mkdir(wd)
        return self.files

    def update(self):
        self.install()

    def install_scripts(self):
        # data for all scripts
        scripts = [
            # bin/code-analysis
            {'bin': (self.name,
                     self.__module__,
                     'code_analysis'), },
            # flake8
            {'bin': 'flake8',
             'arguments': False, },
            # bin/code-analysis-flake8
            {'suffix': 'flake8', },
            # bin/code-analysis-jshint
            {'suffix': 'jshint', },
            # bin/code-analysis-jshint
            {'suffix': 'jscs', },
            # bin/code-analysis-csslint
            {'suffix': 'csslint', },
            # bin/code-analysis-zptlint
            {'suffix': 'zptlint', },
            # bin/code-analysis-deprecated-aliases
            {'suffix': 'deprecated-aliases', },
            # bin/code-analysis-utf8-header
            {'suffix': 'utf8-header', },
            # bin/code-analysis-clean-lines
            {'suffix': 'clean-lines', },
            # bin/code-analysis-prefer-single-quotes
            {'suffix': 'prefer-single-quotes', },
            # bin/code-analysis-pep3101
            {'suffix': 'pep3101', },
            # bin/code-analysis-imports
            {'suffix': 'imports', },
            # bin/code-analysis-debug-statements
            {'suffix': 'debug-statements', },
            # bin/code-analysis-find-untranslated
            {'suffix': 'find-untranslated', },
            # bin/code-analysis-rst-spelling
            {'suffix': 'rst-spelling', },
        ]

        eggs = self.egg.working_set()[1]
        python_buildout = self.buildout['buildout']['python']
        python = self.buildout[python_buildout]['executable']
        directory = self.buildout['buildout']['bin-directory']
        arguments = self.options.__repr__()

        for script in scripts:
            cmd = None
            if 'suffix' in script:
                suffix = script['suffix']
                py_method = suffix.replace('-', '_')

                cmd = ('{0}-{1}'.format(self.name, suffix),
                       self.__module__,
                       'code_analysis_{0}'.format(py_method), )
            elif 'bin' in script:
                cmd = script['bin']
            else:
                raise ValueError('Error trying to install a script. Either'
                                 '"bin" or "suffix" are required.')

            if 'arguments' in script and not script['arguments']:
                zc.buildout.easy_install.scripts(
                    [cmd],
                    eggs,
                    python,
                    directory,
                )
            else:
                zc.buildout.easy_install.scripts(
                    [cmd],
                    eggs,
                    python,
                    directory,
                    arguments=arguments,
                )

    def install_pre_commit_hook(self):
        git_hooks_directory = self.buildout['buildout']['directory'] + \
            '/.git/hooks'
        if not os.path.exists(git_hooks_directory):
            print('Unable to create git pre-commit hook, '
                  'this does not seem to be a git repository.')
            return
        with open(git_hooks_directory + '/pre-commit', 'w') as output_file:
            output_file.write('#!/bin/bash\nbin/code-analysis')
        subprocess.call([
            'chmod',
            '775',
            git_hooks_directory + '/pre-commit',
        ])
        print('Install Git pre-commit hook.')

    def uninstall_pre_commit_hook(self):
        git_hooks_directory = self.buildout['buildout']['directory'] + \
            '/.git/hooks'
        try:
            os.remove(git_hooks_directory + '/pre-commit')
        except OSError:
            pass
        print('Uninstall Git pre-commit hook.')


def code_analysis(options):
    checks = [
        ['flake8', code_analysis_flake8],
        ['jshint', code_analysis_jshint],
        ['jscs', code_analysis_jscs],
        ['csslint', code_analysis_csslint],
        ['zptlint', code_analysis_zptlint],
        ['deprecated-aliases', code_analysis_deprecated_aliases],
        ['utf8-header', code_analysis_utf8_header],
        ['clean-lines', code_analysis_clean_lines],
        ['prefer-single-quotes', code_analysis_prefer_single_quotes],
        ['pep3101', code_analysis_pep3101],
        ['imports', code_analysis_imports],
        ['debug-statements', code_analysis_debug_statements],
        ['find-untranslated', code_analysis_find_untranslated],
        ['rst-spelling', code_analysis_rst_spelling],
    ]
    status_codes = []
    for option, check in checks:
        if option in options and options[option] != 'False':
            status_codes.append(check(options))

    # Check all status codes and return with exit code 1 if one of the code
    # analysis steps did not return True
    if options['return-status-codes'] != 'False':
        for status_code in status_codes:
            if not status_code:
                print('The command "bin/code-analysis" exited with 1.')
                exit(1)
        print('The command "bin/code-analysis" exited with 0.')
        exit(0)
